### PR TITLE
Case 22075: Controller JSDoc update

### PR DESCRIPTION
**Pull request by ctrlaltdavid**
Namepaces:
- Controller

Classes:
- MappingObject
- RouteObject

Primary Manuscript case: https://highfidelity.manuscript.com/f/cases/22075/JSDoc-Controller

Other Manuscript cases addressed:
- https://highfidelity.manuscript.com/f/cases/16213/RouteObject-when-filter-doesn-t-work-with-logical-NOT-values
- https://highfidelity.manuscript.com/f/cases/17029/Typo-in-Controller-API-Documentation
- https://highfidelity.manuscript.com/f/cases/21735/Missing-Documentation-of-Application-Menu-Button-on-Vive#BugEvent.148851

Note: No executable code changes.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18665/HighFidelity-Beta-PR15362-59533d0.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18665/HighFidelity-Quest-Beta-PR15362-59533d0.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18665/HighFidelity-Beta-PR15362-59533d0.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18665/HighFidelity-Beta-Interface-PR15362-59533d0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18665/HighFidelity-Beta-PR15362-59533d0.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18665/HighFidelity-Beta-Interface-PR15362-59533d0.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19043/HighFidelity-Beta-PR15362-407ce74.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19043/HighFidelity-Beta-Interface-PR15362-407ce74.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19043/HighFidelity-Beta-PR15362-407ce74.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19043/HighFidelity-Quest-Beta-PR15362-407ce74.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19043/HighFidelity-Beta-PR15362-407ce74.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19043/HighFidelity-Beta-Interface-PR15362-407ce74.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19084/HighFidelity-Beta-PR15362-3032b58.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19084/HighFidelity-Beta-Interface-PR15362-3032b58.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19084/HighFidelity-Beta-PR15362-3032b58.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19084/HighFidelity-Beta-Interface-PR15362-3032b58.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19084/HighFidelity-Beta-PR15362-3032b58.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19084/HighFidelity-Quest-Beta-PR15362-3032b58.apk)

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -664,9 +664,10 @@ class ApplicationMeshProvider : public scriptable::ModelProviderFactory  {
 
 /**jsdoc
```
Comment (nimisha20) [inline interface/src/Application.cpp]: I couldn't find the MappingObject content in the files changed, so I'm going to add my comment here. 

The links to Controller.Standard and Controller. Hardware just goes to the Controller API page.

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -664,9 +664,10 @@ class ApplicationMeshProvider : public scriptable::ModelProviderFactory  {
...
  * <p>The <code>Controller.Hardware.Application</code> object has properties representing Interface's state. The property
- * values are integer IDs, uniquely identifying each output. <em>Read-only.</em> These can be mapped to actions or functions or
- * <code>Controller.Standard</code> items in a {@link RouteObject} mapping (e.g., using the {@link RouteObject#when} method).
- * Each data value is either <code>1.0</code> for "true" or <code>0.0</code> for "false".</p>
+ * values are integer IDs, uniquely identifying each output. <em>Read-only.</em></p>
+ * <p>These states can be mapped to actions or functions or <code>Controller.Standard</code> items in a {@link RouteObject} 
+ * mapping (e.g., using the {@link RouteObject#when} method). Each data value is either <code>1.0</code> for "true" or
```
Comment (nimisha20) [inline interface/src/Application.cpp]: I couldn't find this object - MappingJSONRoute. Here are some comments about its properties:
Changes requested:
from: The name of a Controller.Hardware property name or an axis made from them. If it's a property name, the leading "Controller.Hardware." can be omitted.
peek: If true, then peeking is enabled per ... (added a comma)
debug: If true, then debug is enabled per... (added a comma)
when: This summary has boolean operations in incorrect formatting. Could you also rephrase this a bit more simply?

File: libraries/controllers/src/controllers/StandardController.cpp [hunk 3]
```diff
@@ -30,17 +30,16 @@ void StandardController::focusOutEvent() {
...
- * and the range <code>-1.0 &ndash; 1.0</code> for bidirectional scales.</p>
- *
- * <p>Each hardware device has a mapping from its outputs to <code>Controller.Standard</code> items, specified in a JSON file.
- * For example, <a href="https://github.com/highfidelity/hifi/blob/master/interface/resources/controllers/leapmotion.json">
- * leapmotion.json</a> and
- * <a href="https://github.com/highfidelity/hifi/blob/master/interface/resources/controllers/vive.json">vive.json</a>.</p>
+ * identifying each output. <em>Read-only.</em></p>
+ * <p>These outputs can be mapped to actions or functions in a {@link RouteObject} mapping. The data value provided by each 
+ * control is either a number or a {@link Pose}. Numbers are typically normalized to <code>0.0</code> or <code>1.0</code> for 
+ * button states, the range <code>0.0 &ndash; 1.0</code> for unidirectional scales, and the range <code>-1.0 &ndash; 1.0</code>
```
Comment (nimisha20) [inline libraries/controllers/src/controllers/StandardController.cpp]: Check if the range values match the formatting with other APIs

File: libraries/controllers/src/controllers/ScriptingInterface.h [hunk 4]
```diff
@@ -212,7 +212,7 @@ namespace controller {
          * @function Controller.triggerHapticPulse
          * @param {number} strength - The strength of the haptic pulse, <code>0.0</code> &ndash; <code>1.0</code>.
```
Comment (nimisha20) [inline libraries/controllers/src/controllers/ScriptingInterface.h]:
```suggestion
         * @param {number} strength - The strength of the haptic pulse, range <code>0.0</code> &ndash; <code>1.0</code>.
```

File: libraries/controllers/src/controllers/ScriptingInterface.h [hunk 5]
```diff
@@ -225,7 +225,7 @@ namespace controller {
          * Triggers a 250ms haptic pulse on connected and enabled devices that have the capability.
          * @function Controller.triggerShortHapticPulse
          * @param {number} strength - The strength of the haptic pulse, <code>0.0</code> &ndash; <code>1.0</code>.
```
Comment (nimisha20) [inline libraries/controllers/src/controllers/ScriptingInterface.h]:
```suggestion
         * @param {number} strength - The strength of the haptic pulse, range <code>0.0</code> &ndash; <code>1.0</code>.
```

File: libraries/controllers/src/controllers/ScriptingInterface.h [hunk 6]
```diff
@@ -235,7 +235,7 @@ namespace controller {
          * @param {number} deviceID - The ID of the device to trigger the haptic pulse on.
          * @param {number} strength - The strength of the haptic pulse, <code>0.0</code> &ndash; <code>1.0</code>.
```
Comment (nimisha20) [inline libraries/controllers/src/controllers/ScriptingInterface.h]:
```suggestion
         * @param {number} strength - The strength of the haptic pulse, range <code>0.0</code> &ndash; <code>1.0</code>.
```

File: libraries/controllers/src/controllers/ScriptingInterface.h [hunk 7]
```diff
@@ -251,18 +251,18 @@ namespace controller {
          * @function Controller.triggerShortHapticPulseOnDevice
          * @param {number} deviceID - The ID of the device to trigger the haptic pulse on.
          * @param {number} strength - The strength of the haptic pulse, <code>0.0</code> &ndash; <code>1.0</code>.
```
Comment (nimisha20) [inline libraries/controllers/src/controllers/ScriptingInterface.h]:
```suggestion
         * @param {number} strength - The strength of the haptic pulse, range <code>0.0</code> &ndash; <code>1.0</code>.
```

File: interface/src/Application.cpp (hunk 1 above)
Comment (ctrlaltdavid) [reply interface/src/Application.cpp]: Links to Controller.Standard etc.: These (e.g., the links at the top of the MappingOejct doc page) work for me (Edge and Internet Explorer). The links work for me, both in my PR and in the existing API docs, https://apidocs.highfidelity.com/MappingObject.html … What browser(s) have you tried the links with?

Comment (ctrlaltdavid) [reply interface/src/Application.cpp]: "Show / hide the tablet" changed to "Show/hide the tablet".

File: libraries/input-plugins/src/input-plugins/KeyboardMouseDevice.cpp [hunk 8]
```diff
@@ -274,9 +275,13 @@ controller::Input KeyboardMouseDevice::InputDevice::makeInput(KeyboardMouseDevic
  *     <tr><td><code>MouseWheelLeft</code></td><td>number</td><td>number</td><td>The mouse wheel rotated left. The data value
```
[the remark this answers is not in the document]
Comment (ctrlaltdavid) [reply libraries/input-plugins/src/input-plugins/KeyboardMouseDevice.cpp]: First two items: see previous comment.

MouseWheelRight: fixed.

File: interface/src/scripting/ControllerScriptingInterface.h [hunk 9]
```diff
@@ -281,15 +295,15 @@ public slots:
     virtual void releaseKeyEvents(const KeyEvent& event);
 
     /**jsdoc
-     * Disable default Interface actions for a joystick.
+     * Disables default Interface actions for a joystick.
      * @function Controller.captureJoystick
      * @param {number} joystickID - The integer ID of the joystick.
      * @deprecated This function is deprecated and will be removed. It no longer has any effect.
```
[the remark this answers is not in the document]
Comment (ctrlaltdavid) [reply interface/src/scripting/ControllerScriptingInterface.h]: Hmmm. Some deprecated items work and some, such as this one, don't. … Just want to double-check that you do want to remove, "It no longer has any effect".

Note: If remove it here, also remove it in Controller.releaseJoystick.

File: interface/src/scripting/ControllerScriptingInterface.h [hunk 10]
```diff
@@ -78,58 +85,65 @@ class ScriptEngine;
...
  * </ul>
  *
- * <p>Controller and Action Values</p>
+ * <h4>Controller and Action Values</h4>
+ * <p>Get the current value of controller outputs and actions.</p>
  * <ul>
  *   <li>{@link Controller.getValue|getValue}</li>
  *   <li>{@link Controller.getAxisValue|getAxisValue}</li>
  *   <li>{@link Controller.getPoseValue|getPoseValue}</li>
  *   <li>{@link Controller.getActionValue|getActionValue}</li>
```
[the remark this answers is not in the document]
Comment (ctrlaltdavid) [reply interface/src/scripting/ControllerScriptingInterface.h]: Some double-check as previous comment.

File: interface/src/Application.cpp (hunk 2 above)
Comment (ctrlaltdavid) [reply interface/src/Application.cpp]: "from": changed to ... from: The name of a Controller.Hardware property or an axis made from them. If it's a property name, the leading "Controller.Hardware." can be omitted. ... I.e., "name" deleted before first "property".
"peek": comma added.
"debug": comma added.
"when": I don't see what incorrect formatting you're referring to. Rephrased.

File: libraries/controllers/src/controllers/StandardController.cpp (hunk 3 above)
Comment (ctrlaltdavid) [reply libraries/controllers/src/controllers/StandardController.cpp]: Fixed.

File: libraries/controllers/src/controllers/ScriptingInterface.h (hunk 4 above)
Comment (ctrlaltdavid) [reply libraries/controllers/src/controllers/ScriptingInterface.h]: Done.

File: libraries/controllers/src/controllers/ScriptingInterface.h (hunk 5 above)
Comment (ctrlaltdavid) [reply libraries/controllers/src/controllers/ScriptingInterface.h]: Done.

File: libraries/controllers/src/controllers/ScriptingInterface.h (hunk 6 above)
Comment (ctrlaltdavid) [reply libraries/controllers/src/controllers/ScriptingInterface.h]: Done.

File: libraries/controllers/src/controllers/ScriptingInterface.h (hunk 7 above)
Comment (ctrlaltdavid) [reply libraries/controllers/src/controllers/ScriptingInterface.h]: Done.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19163/HighFidelity-Beta-PR15362-e217909.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/19163/HighFidelity-Beta-Interface-PR15362-e217909.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19163/HighFidelity-Beta-PR15362-e217909.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/19163/HighFidelity-Quest-Beta-PR15362-e217909.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19163/HighFidelity-Beta-PR15362-e217909.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/19163/HighFidelity-Beta-Interface-PR15362-e217909.exe

File: interface/src/Application.cpp (hunk 1 above)
Comment (nimisha20) [reply interface/src/Application.cpp]: I'm using Chrome but I was just having cache issues. Just tested it again, the links are working!

File: interface/src/scripting/ControllerScriptingInterface.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (nimisha20) [reply interface/src/scripting/ControllerScriptingInterface.h]: If that's the case, let's keep it there.

[the remark this answers is not in the document]
Comment (nimisha20) [reply interface/src/scripting/ControllerScriptingInterface.h]: Let's leave it there if that's the case.

**Pull request merged**